wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -0,0 +1,247 @@
+use {
+    crate::solana::wen_restart_proto::{
+        LastVotedForkSlotsAggregateRecord, LastVotedForkSlotsRecord,
+    },
+    log::*,
+    solana_gossip::crds_value::RestartLastVotedForkSlots,
+    solana_runtime::epoch_stakes::EpochStakes,
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::collections::{HashMap, HashSet},
+};
+
+pub struct LastVotedForkSlotsAggregate {
+    root_slot: Slot,
+    repair_threshold: f64,
+    // TODO(wen): using local root's EpochStakes, need to fix if crossing Epoch boundary.
+    epoch_stakes: EpochStakes,
+    last_voted_fork_slots: HashMap<Pubkey, RestartLastVotedForkSlots>,
+    slots_stake_map: HashMap<Slot, u64>,
+    active_peers: HashSet<Pubkey>,
+    slots_to_repair: HashSet<Slot>,
+}
+
+pub struct LastVotedForkSlotsAggregateResult {
+    pub slots_to_repair: Vec<Slot>,
+    pub active_percent: f64, /* 0 ~ 100.0 */
+}
+
+impl LastVotedForkSlotsAggregate {
+    pub(crate) fn new(
+        root_slot: Slot,
+        repair_threshold: f64,
+        epoch_stakes: &EpochStakes,
+        last_voted_fork_slots: &Vec<Slot>,
+        my_pubkey: &Pubkey,
+    ) -> Self {
+        let mut active_peers = HashSet::new();
+        let my_stake = Self::validator_stake(epoch_stakes, my_pubkey);
+        active_peers.insert(*my_pubkey);
+        let mut slots_stake_map = HashMap::new();
+        for slot in last_voted_fork_slots {
+            if slot <= &root_slot {
+                break;
```

Comment:
Yes it's sorted in Gossip code.

Reply:
The problem is that if it's not sorted, assertion doesn't help you because you exited early already. I'm letting it loop on all slots, a loop on vec<u64> should be really fast.